libs/feature/editor/src/lib/components/record-form/form-field/form-field-contacts-for-resource/form-field-contacts-for-resource.component.ts

```diff
@@ -0,0 +1,220 @@
+import { UserModel } from '@geonetwork-ui/common/domain/model/user'
+import { PlatformServiceInterface } from '@geonetwork-ui/common/domain/platform.service.interface'
+import { OrganizationsServiceInterface } from '@geonetwork-ui/common/domain/organizations.service.interface'
+import { ContactCardComponent } from '../../../contact-card/contact-card.component'
+import {
+  DynamicElement,
+  SortableListComponent,
+} from '@geonetwork-ui/ui/elements'
+
+@Component({
+  selector: 'gn-ui-form-field-contacts-for-resource',
+  templateUrl: './form-field-contacts-for-resource.component.html',
+  styleUrls: ['./form-field-contacts-for-resource.component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  standalone: true,
+  imports: [
+    DropdownSelectorComponent,
+    UiInputsModule,
+    CommonModule,
+    UiWidgetsModule,
+    AutocompleteComponent,
+    TranslateModule,
+    ContactCardComponent,
+    SortableListComponent,
+  ],
+})
+export class FormFieldContactsForResourceComponent
+  implements OnInit, OnChanges
+{
+  @Input() control: FormControl<Individual[]>
+
+  allUsers$: Observable<UserModel[]>
+
+  rolesToPick: string[] = [
+    'resource_provider',
+    'custodian',
+    'owner',
+    'point_of_contact',
+    'author',
+  ]
+
+  rolesToDisplay = []
+
+  allOrganizations: Map<string, Organization> = new Map()
+
+  addOptions: Array<{ buttonLabel: string; eventName: string }> = []
+
+  constructor(
+    private platformServiceInterface: PlatformServiceInterface,
+    private organizationsServiceInterface: OrganizationsServiceInterface,
+    private changeDetectorRef: ChangeDetectorRef
+  ) {
+    this.allUsers$ = this.platformServiceInterface.getUsers()
+  }
+
+  ngOnInit(): void {
+    this.organizationsServiceInterface.organisations$.subscribe(
+      (organizations) => {
+        this.allOrganizations = new Map<string, Organization>(
+          organizations.map((organization) => [organization.name, organization])
+        )
+        this.changeDetectorRef.markForCheck()
+      }
+    )
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    const contactsForResource = changes['control']
+
+    if (
+      contactsForResource &&
+      contactsForResource.currentValue !== contactsForResource.previousValue
+    ) {
+      const rolesToAdd = contactsForResource.currentValue.value.map(
+        (contact: Individual) => contact.role
+      )
+
+      console.log(contactsForResource.currentValue)
+
+      rolesToAdd.forEach((role: string) => {
+        if (!this.rolesToDisplay.includes(role)) {
+          this.rolesToDisplay.push(role)
+        }
+      })
+
+      this.filterRolesToPick()
+
+      this.changeDetectorRef.markForCheck()
+    }
+  }
+
+  addRoleToDisplay(roleToAdd: string) {
+    this.rolesToDisplay.push(roleToAdd)
+    this.filterRolesToPick()
+  }
+
+  filterRolesToPick() {
+    this.rolesToPick = this.rolesToPick.filter(
+      (role) => !this.rolesToDisplay.includes(role)
+    )
+  }
```

Comment:
To add to the comment from @Angi-Kinas , I think you can even do this in a more straightforward way with getters:

```ts
const availableRoles = [
    'resource_provider',
    'custodian',
    'owner',
    'point_of_contact',
    'author',
  ]

get rolesToPick() {
  return availableRoles.filter(
      (role) => !this.control.value.any(contact => contact.role === role)
    )
}

get rolesToDisplay() {
  return this.control.value.reduce((prev, curr) => prev.indexOf(curr.role) > -1 ? prev : [...prev, curr.role], [])
}
```

(untested!)

Reply:
I think you meant some not any.
But anyway, I agree that this is a one liner right, but I don't think this is more readable and more importantly the list of role to display is not only the list of role coming from the input control, it should also contain the ones that the user just clicked !!  So I have to manage another array.

Reply:
Set is indeed a better way to remove duplicates in an array :) I see you're not going with the getter-based approach; it works the way it is but please keep in mind that using getters for computing values in an immediate way instead of maintaining a local intermediary state (in that case the `contactsForRessourceByRole` and `contactsAsDynElemByRole` variables) gives better code readability and is less error-prone because the intermediary state does not need to be kept up-to-date manually